src/gls/server.c3

```diff
@@ -0,0 +1,763 @@
+    }
+
+    let line : Optional<JSON_Node> = location:data().map:at(sv("line"));
+    if !line:has_value() {
+        print(sv("'line' not found"));
+        json:data():deallocate();
+        return;
+    }
+
+    let column : Optional<JSON_Node> = location:data().map:at(sv("column"));
+    if !column:has_value() {
+        print(sv("'column' not found"));
+        json:data():deallocate();
+        return;
+    }
+
+    mut cursor : Cursor = { line:data().value:data():get_iter(), &mut state.allocator };
+    let line_number : Optional<int> = cursor:parse_natural_number();
+
+    if !line_number:has_value() {
+        print(sv("Couldn't parse line number"));
+        json:data():deallocate();
+        return;
+    }
+
+    cursor.iter = column:data().value:data():get_iter();
+    let column_number : Optional<int> = cursor:parse_natural_number();
+
+    if !column_number:has_value() {
+        print(sv("Couldn't parse column number"));
+        json:data():deallocate();
+        return;
+    }
+
+    printf(sv("message: %, line: %"), message:data().value:data(), line:data().value:data());
+    let resp : StringView = sv("{\"jsonrpc\":\"2.0\",\"id\":%,\"result\":{\"kind\":\"full\", \"items\": [{ \"range\": { \"start\": { \"line\": %, \"character\": % }, \"end\" : { \"line\": %, \"character\": % } }, \"message\": \"%\" }]}}");
+    mut ss : StringStream;
+    ss.buffer_pointer = 0;
+    ss:writef(resp, req.id:data(), line_number:data() - 1, column_number:data() - 1, line_number:data() - 1, column_number:data() - 1, message:data().value:data());
+    print(ss:as_sv());
+    out:writef(sv("Content-Length: %\r\n"), ss:length());
+    out:write(sv("\r\n"));
+    out:write(ss:as_sv());
+    out:flush();
+
+    json:data():deallocate();
+}
+
+function handle_text_change : (
+    state : State mut&, req : RPCNotification, out : BufferedFile mut&
+) -> void = {
+    printf(sv("Received textDocument/didChange notification"));
+
+    let text_document : Optional<JSON_Node> = req.params.map:at(sv("textDocument"));
+    if !text_document:has_value() {
+        print(sv("'textDocument' not found"));
+        return;
+    }
+
+    let uri : Optional<JSON_Node> = text_document:data().map:at(sv("uri"));
+    if !uri:has_value() {
+        print(sv("'uri' not found"));
+        return;
+    }
+
+    let version : Optional<JSON_Node> = text_document:data().map:at(sv("version"));
+    if !version:has_value() {
+        print(sv("'version' not found"));
+        return;
+    }
+
+    let content_changes : Optional<JSON_Node> = req.params.map:at(sv("contentChanges"));
+    if !content_changes:has_value() {
+        print(sv("'contentChanges' not found"));
+        return;
+    }
+
+    if content_changes:data().children:length() == 0 {
+        print(sv("'contentChanges' is empty"));
+        return;
+    }
+
+    let text_wrapper : JSON_Node = content_changes:data().children:get(0);
+    let text : Optional<JSON_Node> = text_wrapper.map:at(sv("text"));
```

Comment:
Ah, this took a bit of reasoning... But this always sends over the full document because you replied to initialization with `textDocumentSync.Full`. Cool